lib/pmap.dart

```diff
@@ -1,109 +1,178 @@
-    {int parallel = 1}) {
-  List<_ProcessorIsolate<T, U>> isolates = [];
-  StreamController<U> controller = StreamController<U>();
-  Iterator<T> it = list.iterator;
-  int sendCount = 0;
-  int receiveCount = 0;
-  List<_Enumerated<U>> buffer = [];
-  for (int i = 0; i < parallel; ++i) {
-    _ProcessorIsolate<T, U> isolate = _ProcessorIsolate<T, U>();
-    isolate.processor.mapper = mapper;
-    isolates.add(isolate);
-    isolate.spawn().then((x) async {
-      isolate.receivePort.listen((dynamic result) {
-        if (isolate.sendPort == null) {
-          isolate.sendPort = result;
-          isolate.sendPort.send(isolate.processor);
-          if (it.moveNext()) {
-            isolate.sendPort.send(_Enumerated(index:sendCount++, value:it.current));
-          } else {
-            isolate.completer.complete();
-          }
-        } else {
-          _Enumerated<U> enumeratedResult = result;
-          if (enumeratedResult.index == receiveCount) {
-            controller.add(enumeratedResult.value);
-            receiveCount++;
+/// If preserving the order of the [iterable] is not required,
+/// the [preserveOrder] setting can be set to `false`, reducing memory and
+/// returning results as soon as available.
+///
+/// Note: [mapper] must be a static method or a top-level function.
+Stream<U> pmap<T, U>(
+  Iterable<T> iterable,
+  U Function(T input) mapper, {
+  int parallel = 1,
+  bool preserveOrder = true,
+}) {
+  assert(
+    parallel > 0,
+    'There need to be at least one worker, but got $parallel',
+  );
+
+  final controller = StreamController<U>();
+
+  // If the iterable was a list, the `currentIterableIndex` would be the
+  // index of the element in `iterable.iterator`.
+  var currentIterableIndex = -1;
+  final it = iterable.iterator;
+  bool nextInputElement() {
+    final hasNext = it.moveNext();
+    if (hasNext) currentIterableIndex++;
+    return hasNext;
+  }
+
+// The number of already emitted elements
+  var nextPublishIndex = 0;
+  void publishElement(U result) {
+    controller.add(result);
+    nextPublishIndex++;
+  }
+
+// Did we emit a value for all elements gotten from iterable.
+  var allElementsPublished = false;
+
+  final buffer = <int, U>{};
+  final isolates = List.generate(
+    parallel,
+    (_) async {
+      final isolate = await _ProcessorIsolate.spawn(mapper);
+      isolate.receivePort.listen(
+        (dynamic result) {
+          if (isolate.sendPort == null) {
+            isolate.sendPort = result as SendPort;
           } else {
-            buffer.add(enumeratedResult);
-            int index = buffer.indexWhere((x) => x.index == receiveCount);
-            while (index >= 0) {
-              controller.add(buffer.elementAt(index).value);
-              buffer.removeAt(index);
-              receiveCount++;
-              index = buffer.indexWhere((x) => x.index == receiveCount);
+            final enumeratedResult = result as MapEntry<int, U>;
+            if (preserveOrder) {
```

Comment:
nit: we could get better performance if we pulled out this boolean check so that it doesn't happen at each event.

Reply:
I looked into this and I failed to do that without plain duplicating all the code in which case I would argue that the overhead is not worth this small gain.

If you have a suggestion on doing this more elegantly I am all ears!

What I did improve though, is making the `buffer` `late` so it only gets initialised in case it is used (my first time benefiting from the new keyword 🎉).